app/serializers/osf-serializer.ts

```diff
@@ -147,10 +152,30 @@ export default class OsfSerializer extends JSONAPISerializer {
                     delete serialized.data.attributes![attribute];
                 }
             }
+
+            // The following is a bit of a hack as it relies on currently private information.
+            // When https://github.com/emberjs/data/pull/8131 is merged, we should have public access
+            // to this data from a new import location.
             if (serialized.data.relationships) {
+                const graph = peekGraph(this.store);
                 for (const key of Object.keys(serialized.data.relationships)) {
-                    const rel = serialized.data.relationships[key];
-                    if (rel === null) {
+                    const rel = graph.get(recordIdentifierFor(snapshot.record), camelize(key));
+                    let isClean = true;
+                    if (rel.definition.kind === 'belongsTo') {
+                        isClean = rel.localState === rel.remoteState;
+                    } else if (rel.definition.kind === 'hasMany') {
+                        if (rel.canonicalState.length !== rel.currentState.length) {
+                            isClean = false;
+                        } else {
+                            for (const stateKey in rel.canonicalState) {
+                                if(rel.canonicalState[stateKey] !== rel.currentState[stateKey]) {
+                                    isClean = false;
+                                    break;
+                                }
+                            }
+                        }
+                    }
```

Comment:
The logic seems straightforward and correct: If it is a `belongsTo` relationship, we compare the `localState` and `remoteState` to see if they are the same. If it is a `hasMany` relationship, we first compare the length `canonicalState` and `currentState`, and only compare each entry.

My only concern is with the comparisons themselves. I don't know what `rel.localState` and `rel.remoteState` return. If they return anything but primitives, then the comparison would only work if ember-data is designed in a way that both operands returns the same exact object when the relationship is clean.  The same thing with `canonicalState[stateKey]` and `currentState[stateKey]`.

Reply:
I believe ember data is constructed in such a way as to make those comparisons work. I think they have objects with toString methods on them to facilitate comparisons. 

I can also say that when I messed up some of the logic during development, it failed about 11 tests, and fixing that caused those tests to pass, so I thiiiink everything is okay at the moment and will probably fail tests in the future if ember data breaks something.